File: ChurnReduction.R
#remove all the objects stored
rm(list=ls())
#set current working directory
setwd("C:/Users/SHRAVYA/Desktop/edwisor/project 2")
#install packages
install.packages(c("ggplot2", "corrgram", "DMwR", "caret", "randomForest", "unbalanced", "C50", "dummies", "e1071", "Information", "MASS", "rpart", "gbm", "ROSE", "sampling", "DataCombine", "inTrees"))
library(readxl)     # Super simple excel reader
library(mice)       # missing values imputation
library(naniar)     # visualize missing values
library(dplyr)
library(corrplot)
library(ggplot2)
library(tidyverse)
library(randomForest)
library(caret)
library(data.table)
library(Boruta)
library(rpart)
## Read the data
Train_data = read.csv("Train_data.csv", header = T, na.strings = c(" ", "", "NA"))
Test_data = read.csv("Test_data.csv", header = T, na.strings = c(" ", "", "NA"))
#############################exploratory data analysis##############################
#studying the structure of the given dataset
str(Train_data)
#As we can see, few variables have a wrongly placed datatype.
#The variable , phone.number is actually meant to be a continuoos variable but it is analysed as a factor variable with 3333 levels. hence it has to be altered
#The variable area.code has to be a factor variable.
Train_data$phone.number  = as.numeric(Train_data$phone.number )
Test_data$phone.number  = as.numeric(Test_data$phone.number )
Train_data$area.code  = as.factor(Train_data$area.code )
Test_data$area.code  = as.factor(Test_data$area.code )
str(Train_data)
#################################missing value analysis###########################
missing_val = data.frame(apply(Train_data,2,function(x){sum(is.na(x))}))
missing_val$Columns = row.names(missing_val)
names(missing_val)[1] =  "Missing_percentage"
missing_val$Missing_percentage = (missing_val$Missing_percentage/nrow(Train_data)) * 100
missing_val = missing_val[order(-missing_val$Missing_percentage),]
row.names(missing_val) = NULL
missing_val = missing_val[,c(2,1)]
write.csv(missing_val, "Missing_perc.csv", row.names = F)
sum(is.na(Train_data))
sum(is.na(Test_data))
#as we can see, there are no missing values. Hence no missing value imputation is requirede.calls, Churn)
############################################variable analysis#########################
library(ggplot2)
library(ggpubr)
theme_set(theme_pubr())
ggplot(Train_data, aes(Churn)) +
  geom_bar(fill = "#0073C2FF") +
  theme_pubclean()
#outlier analysis
numeric_index = sapply(Train_data,is.numeric) #selecting only numeric

numeric_data = Train_data[,numeric_index]

cnames = colnames(numeric_data)
 
 for (i in 1:length(cnames))
 {
assign(paste0("gn",i), ggplot(aes_string(y = (cnames[i]), x = "Churn"), data = subset(Train_data))+ 
           stat_boxplot(geom = "errorbar", width = 0.5) +
            geom_boxplot(outlier.colour="red", fill = "grey" ,outlier.shape=18,
                         outlier.size=1, notch=FALSE) +
            theme(legend.position="bottom")+
            labs(y=cnames[i],x="responded")+
            ggtitle(paste("Box plot of churned for",cnames[i])))
}

# ##Plotting plots together
gridExtra::grid.arrange(gn1,gn5,gn2,ncol=3)
gridExtra::grid.arrange(gn6,gn7,ncol=2)
gridExtra::grid.arrange(gn8,gn9,ncol=2)
#feature selection
################################### Correlation Plot - to check multicolinearity between continous variables
library(corrgram)
corrgram(Train_data[,numeric_index], order = F,
         upper.panel=panel.pie, text.panel=panel.txt, main = "Correlation Plot")
#few independent variables are found to be highly correlated to each other. hence one of each copy is eliminated
## Chi-squared Test of Independence-to check the multicolinearity between categorical variables
factor_index = sapply(Train_data,is.factor)
factor_data = Train_data[,factor_index]
for (i in 1:5)
{
  print(names(factor_data)[i])
  print(chisq.test(table(factor_data$Churn,factor_data[,i])))
}
#the variable area code has a p value of greater than 0.05. Hence we eliminate the variable
## Dimension Reduction
Train_data = subset(Train_data, 
                         select = -c(area.code,total.day.minutes,total.eve.minutes, total.night.minutes, total.intl.minutes))
Test_data = subset(Test_data, 
                    select = -c(area.code,total.day.minutes,total.eve.minutes, total.night.minutes, total.intl.minutes))
###################################model developement###############################
#Clean the environment
library(DataCombine)
rm(list= ls()[!(ls() %in% c('Train_data','Test_data'))])
##Decision tree for classification
#Develop Model on training data
install.packages("C50")
library(C50)
C50_model = C5.0(Churn ~., Train_data, trials = 100, rules = TRUE)
#Summary of DT model
summary(C50_model)
#write rules into disk
write(capture.output(summary(C50_model)), "c50Rules.txt")
#Lets predict for test cases
C50_Predictions = predict(C50_model, Test_data[,-16], type = "class")
##Evaluate the performance of classification model
library(caret)
ConfMatrix_C50 = table(Test_data$Churn, C50_Predictions)
confusionMatrix(ConfMatrix_C50)
#accuracy=95.86%
#false negetive rate=27.67%
########################################Random Forest####################################
library(randomForest)
RF_model = randomForest(Churn ~ ., Train_data, importance = TRUE, ntree = 1000)
#Extract rules fromn random forest
#transform rf object to an inTrees' format
library(RRF)
library(inTrees)
treeList = RF2List(RF_model)  
# 
# #Extract rules
exec = extractRules(treeList, Train_data[,-16])  # R-executable conditions
# #Make rules more readable:
readableRules = presentRules(exec, colnames(Train_data))
# #Get rule metrics
ruleMetric = getRuleMetric(exec, Train_data[,-16], Train_data$Churn)  # get rule metrics
#Presdict test data using random forest model
RF_Predictions = predict(RF_model, Test_data[,-16])

##Evaluate the performance of classification model
ConfMatrix_RF = table(Test_data$Churn, RF_Predictions)
confusionMatrix(ConfMatrix_RF)

#False Negative rate
FNR = FN/FN+TP 

#Accuracy = 94.66
#FNR = 31.25
#Logistic Regression
logit_model = glm(Churn ~ ., data = Train_data, family = "binomial")

#summary of the model
summary(logit_model)

#####################################predict using logistic regression##################################
logit_Predictions = predict(logit_model, newdata = Test_data, type = "response")

#convert prob
logit_Predictions = ifelse(logit_Predictions > 0.5, 1, 0)


##Evaluate the performance of classification model
ConfMatrix_LR = table(Test_data$Churn, logit_Predictions)
confusionMatrix(ConfMatrix_LR)

#False Negative rate
FNR = FN/FN+TP 

#Accuracy: 90.89
#FNR: 67.85
set.seed(4543)
Train.rf <- randomForest(Churn ~ ., data=Train_data, ntree=1000, keep.forest=FALSE,
                          importance=TRUE)
varImpPlot(Train.rf)

File: ChurnReduction.py

# coding: utf-8

# In[5]:


#Load libraries
import os
import pandas as pd
import numpy as np  
import matplotlib.pyplot as plt
from scipy.stats import chi2_contingency
import seaborn as sns
from random import randrange, uniform
from sklearn.cross_validation import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn import linear_model
from sklearn.cross_validation import train_test_split


# In[6]:


#Set working directory
os.chdir("C:/Users/SHRAVYA/Desktop/edwisor/project 2")


# In[7]:


#Load data
Train_data= pd.read_csv("Train_data.csv")
Test_data = pd.read_csv("Test_data.csv")


# In[8]:


#----------------------------------PRE PROCESSING-EXPLORATORY DATA ANALYSIS----------------------------------------------------#
#As we can see, few variables have a wrongly placed datatype.
#The variable , phone.number is actually meant to be a continuoos variable but it is analysed as a factor variable with 3333 levels. hence it has to be altered
#The variable area.code has to be a factor variable.
Train_data['area code']=Train_data['area code'].astype(object)
Test_data['area code']=Test_data['area code'].astype(object)



# In[5]:


#---------------------------------------MISSING VALUE ANALYSIS-----------------------------------------------------------------#
#Create dataframe with missing percentage
missing_val = pd.DataFrame(Train_data.isnull().sum())

#Reset index
missing_val = missing_val.reset_index()

#Rename variable
missing_val = missing_val.rename(columns = {'index': 'Variables', 0: 'Missing_percentage'})

#Calculate percentage
missing_val['Missing_percentage'] = (missing_val['Missing_percentage']/len(Train_data))*100

#descending order
missing_val = missing_val.sort_values('Missing_percentage', ascending = False).reset_index(drop = True)

#save output results 
missing_val.to_csv("Missing_perc.csv", index = False)
Train_data.dtypes
missing_val
Train_data.isnull().sum()
Test_data.isnull().sum()


# In[9]:


#As we can see, no missing values are present. Hence no missing analysis is required.
Train_data.dtypes


# In[7]:


#Plot boxplot to visualize Outliers
get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(Train_data['account length'])


# In[8]:


get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(Train_data['number vmail messages'])


# In[9]:


get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(Train_data['total day minutes'])


# In[10]:


get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(Train_data['total day calls'])


# In[11]:


get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(Train_data['total day charge'])


# In[12]:


get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(Train_data['total eve minutes'])


# In[13]:


get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(Train_data['total eve calls'])


# In[14]:


get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(Train_data['total eve charge'])


# In[15]:


get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(Train_data['total night minutes'])


# In[16]:


get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(Train_data['total night calls'])


# In[17]:


get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(Train_data['total night charge'])


# In[18]:


get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(Train_data['total intl minutes'])


# In[19]:


get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(Train_data['total intl calls'])


# In[20]:


get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(Train_data['total intl charge'])


# In[21]:


get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(Train_data['number customer service calls'])


# In[22]:


#save numeric names
cnames =  ["account length", "number vmail messages", "total day minutes", "total day calls", "total day charge", "total eve minutes", "total eve calls", "total eve charge",
           "total night minutes", "total night calls", "total night charge","total intl minutes","total intl calls","total intl charge","number customer service calls"]


# In[23]:


#Detect and delete outliers from data
for i in cnames:
     print(i)
     q75, q25 = np.percentile(Train_data.loc[:,i], [75 ,25])
     iqr = q75 - q25

     min = q25 - (iqr*1.5)
     max = q75 + (iqr*1.5)
     print(min)
     print(max)
    
     Train_data = Train_data.dropTrain_data[Train_data.loc[:,i] < min].index)
     Train_data = Train_data.drop(Train_data[Train_data.loc[:,i] > max].index)


# In[24]:


##-----------------------------------------------Correlation analysis------------------------------------------------------
#Correlation plot
df_corr = Train_data.loc[:,cnames]


# In[25]:


#Set the width and hieght of the plot
f, ax = plt.subplots(figsize=(7, 5))

#Generate correlation matrix
corr = df_corr.corr()

#Plot using seaborn library
sns.heatmap(corr, mask=np.zeros_like(corr, dtype=np.bool), cmap=sns.diverging_palette(220, 10, as_cmap=True),
            square=True, ax=ax)


# In[26]:


#Chisquare test of independence
#Save categorical variables
cat_names = ["state", "area code", "phone number", "international plan", "voice mail plan"]


# In[27]:


#loop for chi square values
for i in cat_names:
    print(i)
    chi2, p, dof, ex = chi2_contingency(pd.crosstab(Train_data['Churn'], Train_data[i]))
    print(p)


# In[28]:


Train_data.dtypes
Test_data.dtypes


# In[10]:


Test_data = Test_data.drop(['state','area code', 'total day minutes', 'total eve minutes', 'total night minutes', 'total intl minutes','phone number'], axis=1)


# In[11]:


Train_data = Train_data.drop(['state','area code', 'total day minutes', 'total eve minutes', 'total night minutes', 'total intl minutes','phone number'], axis=1)


# In[12]:


Train_data['international plan'] = Train_data['international plan'].astype('category')
Train_data['voice mail plan'] = Train_data['voice mail plan'].astype('category')
Train_data['Churn'] = Train_data['Churn'].astype('category')
Test_data['international plan'] = Test_data['international plan'].astype('category')
Test_data['voice mail plan'] = Test_data['voice mail plan'].astype('category')
Test_data['Churn'] = Test_data['Churn'].astype('category')


# In[13]:


Train_data['international plan'] = (Train_data['international plan'] == ' yes').astype(int)
Train_data['voice mail plan'] = (Train_data['voice mail plan'] == ' yes').astype(int)
Train_data['Churn'] = (Train_data['Churn'] == ' True.').astype(int)
Test_data['international plan'] = (Test_data['international plan'] == ' yes').astype(int)
Test_data['voice mail plan'] = (Test_data['voice mail plan'] == ' yes').astype(int)
Test_data['Churn'] = (Test_data['Churn'] == ' True.').astype(int)


# In[15]:


#Divide data into train and test
X = Train_data.values[:, 0:13]
Y = Train_data.values[:,13]

X_train, X_test, y_train, y_test = train_test_split( X, Y, test_size = 0.2)


# In[35]:


#Decision Tree
C50_model = tree.DecisionTreeClassifier(criterion='entropy').fit(X_train, y_train)

#predict new test cases
C50_Predictions = C50_model.predict(X_test)

#Create dot file to visualise tree  #http://webgraphviz.com/
# dotfile = open("pt.dot", 'w')
# df = tree.export_graphviz(C50_model, out_file=dotfile, feature_names = marketing_train.columns)


# In[36]:


#build confusion matrix
# from sklearn.metrics import confusion_matrix 
# CM = confusion_matrix(y_test, y_pred)
CM = pd.crosstab(y_test, C50_Predictions)

#let us save TP, TN, FP, FN
TN = CM.iloc[0,0]
FN = CM.iloc[1,0]
TP = CM.iloc[1,1]
FP = CM.iloc[0,1]

#check accuracy of model
#accuracy_score(y_test, y_pred)*100
((TP+TN)*100)/(TP+TN+FP+FN)
#accuracy= 90.55%
#False Negative rate=19.0 
(FN*100)/(FN+TP)


# In[37]:


#Random Forest
from sklearn.ensemble import RandomForestClassifier

RF_model = RandomForestClassifier(n_estimators = 20).fit(X_train, y_train)


# In[38]:


RF_Predictions = RF_model.predict(X_test)


# In[39]:


#build confusion matrix
# from sklearn.metrics import confusion_matrix 
# CM = confusion_matrix(y_test, y_pred)
CM = pd.crosstab(y_test, RF_Predictions)

#let us save TP, TN, FP, FN
TN = CM.iloc[0,0]
FN = CM.iloc[1,0]
TP = CM.iloc[1,1]
FP = CM.iloc[0,1]

#check accuracy of model
#accuracy_score(y_test, y_pred)*100
((TP+TN)*100)/(TP+TN+FP+FN)

#False Negative rate 
(FN*100)/(FN+TP)

#Accuracy: 94.75%
FNR: 28.26


# In[46]:


#KNN implementation
from sklearn.neighbors import KNeighborsClassifier

KNN_model = KNeighborsClassifier(n_neighbors = 9).fit(X_train, y_train)


# In[47]:


#predict test cases
KNN_Predictions = KNN_model.predict(X_test)


# In[50]:


#build confusion matrix
CM = pd.crosstab(y_test, KNN_Predictions)

#let us save TP, TN, FP, FN
TN = CM.iloc[0,0]
FN = CM.iloc[1,0]
TP = CM.iloc[1,1]
FP = CM.iloc[0,1]

#check accuracy of model
#accuracy_score(y_test, y_pred)*100
((TP+TN)*100)/(TP+TN+FP+FN)

#False Negative rate 
(FN*100)/(FN+TP)

#Accuracy: 87.10
#FNR: 97


# In[51]:


#Naive Bayes
from sklearn.naive_bayes import GaussianNB

#Naive Bayes implementation
NB_model = GaussianNB().fit(X_train, y_train)


# In[52]:


#predict test cases
NB_Predictions = NB_model.predict(X_test)


# In[56]:


#Build confusion matrix
CM = pd.crosstab(y_test, NB_Predictions)

#let us save TP, TN, FP, FN
TN = CM.iloc[0,0]
FN = CM.iloc[1,0]
TP = CM.iloc[1,1]
FP = CM.iloc[0,1]

#check accuracy of model
#accuracy_score(y_test, y_pred)*100
((TP+TN)*100)/(TP+TN+FP+FN)

#False Negative rate 
#(FN*100)/(FN+TP)

#Accuracy: 86.95
#FNR: 64.36


# In[1]:


#logistic regression
import numpy as np 
import matplotlib.pyplot as plt


# In[16]:


from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
logisticRegr = LogisticRegression()


# In[18]:


logisticRegr.fit(X_train, y_train)


# In[20]:


logisticRegr.predict(X_test[0:13])


# In[21]:


predictions = logisticRegr.predict(X_test)


# In[22]:


score = logisticRegr.score(X_test, y_test)
print(score)
#0.8740629685157422


# In[23]:


import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import metrics


# In[24]:


cm = metrics.confusion_matrix(y_test, predictions)
print(cm)
#FNR = 79.01

